### PR TITLE
Fix DeviceHistogram::Even for mixed float/int levels and sample types.

### LINKED ISSUE #479: cub::DeviceHistogram::HistogramEven: Incorrect result when `LevelT` does not exactly match `SampleT`
Passing into `cub::DeviceHistogram::HistogramEven` bounds of type different from type of samples leads to wrong histogramming. Example constructed from the library code is below. If you change type of `lower_level` and `upper_level` to `float`, all is OK, but `int` as in the example produces wrong histogram [1, 0, 0, 0, 4, 0]. Applicable to 1.16.

``` cxx
#include <cstdlib>
#include <vector>

#include <cub/cub.cuh>   // or equivalently <cub/device/device_histogram.cuh>

// Declare, allocate, and initialize device-accessible pointers for input samples and
// output histogram
int      num_samples = 10;//32 << 20;    // e.g., 10
float*   d_samples;      // e.g., [2.2, 6.1, 7.1, 2.9, 3.5, 0.3, 2.9, 2.1, 6.1, 999.5]
int*     d_histogram;    // e.g., [ -, -, -, -, -, -]
int      num_levels = 7;     // e.g., 7       (seven level boundaries for six bins)
int      lower_level = 0;
int      upper_level = 12;
//float    lower_level = 0.0;    // e.g., 0.0     (lower sample value boundary of lowest bin)
//float    upper_level = 12.0;    // e.g., 12.0    (upper sample value boundary of upper bin)

int main()
{
    std::vector<float> s {2.2, 6.1, 7.1, 2.9, 3.5, 0.3, 2.9, 2.1, 6.1, 999.5};
    std::vector<int> h(num_levels - 1);

    s.resize(num_samples);
    cudaMalloc(&d_samples, num_samples * sizeof(float));
    cudaMalloc(&d_histogram, (num_levels - 1) * sizeof(int));
    cudaMemcpy(d_samples, s.data(), num_samples * sizeof(float), cudaMemcpyHostToDevice);

    // Determine temporary device storage requirements
    void*    d_temp_storage = NULL;
    size_t   temp_storage_bytes = 0;
    cub::DeviceHistogram::HistogramEven(d_temp_storage, temp_storage_bytes,
        d_samples, d_histogram, num_levels, lower_level, upper_level, num_samples);

    // Allocate temporary storage
    cudaMalloc(&d_temp_storage, temp_storage_bytes);

    // Compute histograms
    cub::DeviceHistogram::HistogramEven(d_temp_storage, temp_storage_bytes,
        d_samples, d_histogram, num_levels, lower_level, upper_level, num_samples);

    // d_histogram   <-- [1, 5, 0, 3, 0, 0];

    cudaMemcpy(h.data(), d_histogram, (num_levels - 1) * sizeof(int), cudaMemcpyDeviceToHost);

    for(int a : h)
        std::cout << a << ", ";

    return 0;
}
```

### PR DESCRIPTION
**Pull request by allisonvacanti**
The `ScaleTransform` utility precomputes the reciprocal of the "`sample` -> `bin_idx`" scaling factor for floating point types as an optimization.

Trouble is, the `Init` method checked whether `LevelT` is fp while `BinSelect` checked `SampleT`. This caused the optimization to be incorrectly applied when one of these types is fp but the other is not.

Fixed this bug and simplified the implementation of `ScaleTransform` using `cuda::std::common_type` to consistently apply the optimization when either `LevelT` or `ScaleT` are fp.

Fixes #479 and #489 and adds regression tests for both.

### Breaking Changes:

- NVIDIA/cub#487: Fixed the `DeviceHistogram::HistogramEven` algorithms when the samples and levels are different types (NVIDIA/cub#479) and when both are integral with fractional bin sizes (NVIDIA/cub#489). These fixes introduced the following explicit restrictions:
  - Both `SampleT` and `LevelT` must support common arithmetic operations.
  - `cuda::std::common_type<SampleT, LevelT>` must have a valid definition.
  - The common type must be convertible to `int`.
  - The common type must be trivially copyable.

### CODE REVIEW

**Comment by allisonvacanti** [comment]
gpuCI: NVIDIA/thrust#1697

**Comment by allisonvacanti** [comment]
@Melirius This should fix your issue with `DeviceHistogram`. Let us know if you get a chance to test it out.

**Comment by Melirius** [comment]
It is nice, thanks for your work! However, it does not address another problem https://github.com/NVIDIA/cub/issues/489 that is rooted in the same code.

File: cub/device/dispatch/dispatch_histogram.cuh
Comment (senior-zero) [inline cub/device/dispatch/dispatch_histogram.cuh:244]: I think this is a breaking change. Requirements on types used to be: 
```cpp
template<typename L>
concept Layout = requires(L&& l, L &&r)
{
    requires std::is_trivially_copyable_v<L>;
    { l >= r } -> std::convertible_to<bool>;
    { l < r } -> std::convertible_to<bool>;
    { (l - r) / r } -> std::convertible_to<int>;
};

template<typename S, typename L>
concept Sample = Layout<L> && requires
{
    requires std::is_convertible_v<S, L>;
    requires std::is_copy_constructible_v<S>;
};

template <Layout L, Sample<L> ST>
void histogram(int num_levels, 
               L min,
               L max,
               ST* samples);
```

Now they are something along the following lines:
```cpp
template<typename L, typename R>
concept Common = requires(
    std::common_type_t<L, R> l, 
    std::common_type_t<L, R> r)
{
    requires std::is_convertible_v<int, std::common_type_t<L, R>>;
    typename std::common_type<L, R>::type;
    { l >= r } -> std::convertible_to<bool>;
    { l < r } -> std::convertible_to<bool>;
    { (l - r) / r } -> std::convertible_to<int>;
};

template<typename L, typename CommonT>
concept Layout = requires
{
    requires std::is_trivially_copyable_v<L>;
    requires std::is_convertible_v<L, CommonT>;
};

template<typename S, typename CommonT>
concept Sample = requires
{
    requires std::is_copy_constructible_v<S>;
    requires std::is_convertible_v<S, CommonT>;
};
```

The new constrain `std::is_convertible_v<int, std::common_type_t<L, R>>` is caused by `static_cast<CommonT>(num_output_levels - 1)`. Besides that, a different issue might be found in the following use case:

```cpp
class custom_layout_t
{
  int val{};
public:
  explicit custom_layout_t(int v) : val(v) {}

  operator int() {
    return val;
  }

  // Methods ignored
  custom_layout_t operator-(const custom_layout_t &rhs) { return custom_layout_t{val - rhs.val}; }
  bool operator>=(const custom_layout_t &rhs) { return val >= rhs.val; }
  bool operator<(const custom_layout_t &rhs) { return val < rhs.val; }
  int operator/(const custom_layout_t &rhs) { return val / rhs.val; }
};

int main() {
  custom_layout_t min {0};
  custom_layout_t max {2};

  int samples[2] { 0, 1 };

  int num_levels = 2;

  histogram(num_levels, min, max, samples);
}
```

Before this changes, we'd use `custom_layout_t` methods, whereas now, we'd use integer arithmetic (since the common type is `int`). I think there might be other subtle differences in the requirements and/or behavior. 

My suggestion would be to start with the documentation and provide some constrains on the types. Do we expect the algorithm to work with user types at all?

Comment (senior-zero) [inline cub/device/dispatch/dispatch_histogram.cuh:244]: Member variables are copied on the device, could you add a static assert on `std::is_trivially_copyable<CommonT>`? Since this is going to be a part of 2.0 release, we might  break some code that relies on UB.

**Review by senior-zero — CHANGES_REQUESTED** [review]
Thank you for starting the work on this! It seems that no one looked at the code for a while :smile: 
Regarding the PR, it seems to introduce some substantial changes. I'd like us to consider all implications before merging this PR. For instance, documentation doesn't say much about `SampleT` and `LevelT`. Unlike `ConterT`, `LevelT` isn't even required to be primitive. Let's gather requirements on these types, update documentation / assert these requirements and then have another review.

File: cub/device/dispatch/dispatch_histogram.cuh
Comment (Melirius) [reply cub/device/dispatch/dispatch_histogram.cuh:244]: Integer arithmetics is not a good idea at all, see https://github.com/NVIDIA/cub/issues/489

File: cub/device/dispatch/dispatch_histogram.cuh
Comment (allisonvacanti) [reply cub/device/dispatch/dispatch_histogram.cuh:258]: These are not public APIs. The only API in the Dispatch layer that users should interact with is the `cub::DispatchAlgo::Dispatch(...)` entry point. I'll move these transform classes to be private in the Dispatch class.

File: cub/device/dispatch/dispatch_histogram.cuh
Comment (allisonvacanti) [reply cub/device/dispatch/dispatch_histogram.cuh:260]: Removed.

File: cub/device/dispatch/dispatch_histogram.cuh
Comment (allisonvacanti) [reply cub/device/dispatch/dispatch_histogram.cuh:250]: Good point about windows. Fixing to follow the suggested naming conventions.

File: cub/device/dispatch/dispatch_histogram.cuh
Comment (allisonvacanti) [reply cub/device/dispatch/dispatch_histogram.cuh:287]: Per the draft NVBench conventions, the `this->` pointer is omitted for `m_` prefixed member variables. I will be switching these to use the prefix.

File: cub/device/dispatch/dispatch_histogram.cuh
Comment (allisonvacanti) [reply cub/device/dispatch/dispatch_histogram.cuh:255]: The `Init` pattern is used in the other helper classes for histogram. This PR is already experiencing quite a bit of scope creep, so I'll leave this as is for consistency with the other helpers. This can be done in a future cleanup.

File: cub/device/dispatch/dispatch_histogram.cuh
Comment (allisonvacanti) [reply cub/device/dispatch/dispatch_histogram.cuh:247]: It is not, and it should not be per @griwes:

> `is_floating_point` asks a very specific question, which is: is this a built-in floating point type. `cuda::std::is_floating_point` will never answer "yes" for `__half`, because `__half` does not behave like a built-in floating point type. [...] [Fixing] this would (1) go against what the standard defines it to mean, and (2) would have a potential of breaking code that assumes that `is_floating_point` and `is_class` are never true at the same time.

I've updated this to incorporate your recent changes for `__half`

File: cub/device/dispatch/dispatch_histogram.cuh
Comment (allisonvacanti) [reply cub/device/dispatch/dispatch_histogram.cuh:279]: I agree with the suggestion, but it's not worth refactoring at the current time.

File: cub/device/dispatch/dispatch_histogram.cuh
Comment (allisonvacanti) [reply cub/device/dispatch/dispatch_histogram.cuh:244]: The error is fairly clear, pointing to an invalid specialization of common_type:

```
error: class "cuda::std::__4::common_type<Foo, Bar>" has no member "type"
```

https://www.godbolt.org/z/98rTr7Ez5

File: cub/device/dispatch/dispatch_histogram.cuh
Comment (allisonvacanti) [reply cub/device/dispatch/dispatch_histogram.cuh:291]: Added.

File: cub/device/dispatch/dispatch_histogram.cuh
Comment (allisonvacanti) [reply cub/device/dispatch/dispatch_histogram.cuh:268]: I'd rather do this arithmetic using the native types, and reusing the input variables reduces dependencies to improve ILP.

The compiler *should* optimize out the dependencies in this situation, but just in case...

Comment (allisonvacanti) [reply cub/device/dispatch/dispatch_histogram.cuh:268]: Let's hold off on this until we have the host-assertion machinery in place from your recent debug-level PR.

File: cub/device/dispatch/dispatch_histogram.cuh
Comment (allisonvacanti) [reply cub/device/dispatch/dispatch_histogram.cuh:273]: Same as above.

File: cub/device/dispatch/dispatch_histogram.cuh
Comment (allisonvacanti) [reply cub/device/dispatch/dispatch_histogram.cuh:244]: Added compile time checks for the `common_type` and `is_convertible` requirements and updated documentation to explicitly specify these.

Also updated the logic to fix the round issue reported in #489.

Comment (allisonvacanti) [reply cub/device/dispatch/dispatch_histogram.cuh:244]: Done 👍🏻 Good catch!

File: cub/device/dispatch/dispatch_histogram.cuh
Comment (allisonvacanti) [reply cub/device/dispatch/dispatch_histogram.cuh:258]: Looks like we can't make these private without extensive refactoring:

```
error: A type that is defined inside a class and has private or protected access ("cub::DispatchHistogram<1, 1, int *, int, int, int>::SearchTransform<int *>") cannot be used in the template argument type of a __global__ function template instantiation, unless the class is local to a __device__ or __global__ function
```

File: cub/device/dispatch/dispatch_histogram.cuh
Comment (senior-zero) [inline cub/device/dispatch/dispatch_histogram.cuh:247]: I'm sorry, I've confused something, the `_t` is not defined for C++11 :( 
```suggestion
      using CommonT = typename cuda::std::common_type<LevelT, SampleT>::type;
```

File: cub/device/dispatch/dispatch_histogram.cuh
Comment (senior-zero) [inline cub/device/dispatch/dispatch_histogram.cuh:302]: I don't think this is equivalent to the previous code. In the case of C++11 you are still going to compile the branch above. This leads to compilation error in the case of `__half`, right?

**Review by senior-zero — APPROVED** [review]
Thanks for documenting actual requirements! A few minor fixes below.

File: cub/detail/cpp_compatibility.cuh
Comment (miscco) [inline cub/detail/cpp_compatibility.cuh:23]: Just to be sure, we cannot pass `-Wno-c++17-extensions` to our build flags

Asking for a friend

Comment (allisonvacanti) [reply cub/detail/cpp_compatibility.cuh:23]: No -- Thrust/CUB are header only, so that would require our users to do the same.

File: cub/device/dispatch/dispatch_histogram.cuh
Comment (allisonvacanti) [reply cub/device/dispatch/dispatch_histogram.cuh:303]: Yep -- switching it back to a static dispatch.

**Pull request merged**
